### DISCUSSION TITLE
wireless emulation in multiple servers

### DISCUSSION THREAD

**Discussion by PengfeiM** [coreemu/core#532] [Q&A] [answered]
After installation, I can run wired and wireless(basic and emane) emulation locally. 

And when I config core as follows:(192.168.189.121 is master server's IP)
![core-conf](https://user-images.githubusercontent.com/45191163/99942296-66f5b100-2daa-11eb-881c-04a6fb40f669.jpg)
I can run a wired emulation in multiple servers. 
But when I try to run a emane emulation with all nodes in master server, an error occurs as following:
![emane-local](https://user-images.githubusercontent.com/45191163/99942370-88ef3380-2daa-11eb-851f-bc7890368e19.jpg)
And when I run a emane emulation with nodes in both master and slave servers, an error occurs as following:

![emane-remote](https://user-images.githubusercontent.com/45191163/99942484-b340f100-2daa-11eb-972a-90fb3d003a50.jpg)

Besides, I have a question for distributed emane. 
In documentation of Distribution, it says that I only need to install core in master server. 
![image](https://user-images.githubusercontent.com/45191163/99942634-f3a06f00-2daa-11eb-8173-ea44963de01e.png)

But in documentation of EMANE, it says that each machine that will act as an emulation server needs to have CORE and EMANE installed.
![image](https://user-images.githubusercontent.com/45191163/99942649-f733f600-2daa-11eb-83d4-80ad5286f0f1.png)

So, which instruction should I follow?

**Comment by PengfeiM** [comment]
I check the /tmp/pycore.xxxx/.  path,   find out that ,there is .log file  in master server but no .log file in distributed server.

**Comment by bharnden** [comment]
EMANE being a separate component and is not a default part of CORE. If EMANE is desired, then yes, EMANE is required to be installed on each machine.

**Comment by PengfeiM** [comment]
So, that means the server should install packges as follosing:
**master server**: CORE, EMANE, core_distributed
**distributed server**: EMANE, core_distributed
Am I right?

**Comment by PengfeiM** [comment]
I installed CORE and EMANE in both master server and distributed server, the error `n2-platform.xml didn't exist` disapeared.

But I still get the error bellow:
---
> 240 - ERROR - corehandlers:handle_message - Thread-5: exception while handling message: CoreEventMessage <msgtype = EVENT, flags = 0x0 <>>
>   TYPE: 3**CORE反馈的错误信息**
> 
> Traceback (most recent call last):
>   File "/home/infonet/.cache/pypoetry/virtualenvs/core-bjdTMDRL-py3.6/lib/python3.6/site-packages/emane/shell/controlportclient.py", line 150, in `__init__`
>     `self._sock.connect((hostname,port))`
> ***ConnectionRefusedError: [Errno 111] Connection refused***
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/api/tlv/corehandlers.py", line 543, in `handle_message`
>     `replies = message_handler(message)`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/api/tlv/corehandlers.py", line 1569, in `handle_event_message`
>     `self.session.instantiate()`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/emulator/session.py", line 1184, in `instantiate`
>     `if self.emane.startup() == EmaneState.NOT_READY:`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/emane/emanemanager.py", line 354, in `startup`
>     `self.link_monitor.start()`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/emane/linkmonitor.py", line 193, in `start`
>     `self.initialize()`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/emane/linkmonitor.py", line 206, in `initialize`
>     `client = EmaneClient(address)`
>   File "/home/infonet/core/core-release-7.2.1/daemon/core/emane/linkmonitor.py", line 83, in `__init__`
>     `self.address, DEFAULT_PORT`
>   File "/home/infonet/.cache/pypoetry/virtualenvs/core-bjdTMDRL-py3.6/lib/python3.6/site-packages/emane/shell/controlportclient.py", line 152, in `__init__`
>     `raise ControlPortException("unable to connect to %s:%d" % (hostname,port),True)`
> ***emane.shell.controlportexception.ControlPortException: unable to connect to 192.168.6.2:47000***
---
My configuration is:
---
```
distributed_address = 192.168.6.106
controlnet = 192.168.6.0/24
emane_event_generate = True
```
---
192.168.6.106 is the master server's ip

**Comment by bharnden** [comment, marked as answer]
Don't have the controlnet addresses be the same as your actual host address, that woul dlikely cause a conflict. Set it to something like the default `controlnet = 172.16.0.0/24`.

**Comment by bharnden** [comment]
> So, that means the server should install packges as follosing:
master server: CORE, EMANE, core_distributed
distributed server: EMANE, core_distributed
Am I right?

Master server does not  need core_distributed, since the executables it contains are a part of the normal CORE installation. It is only for distributed nodes.

**Comment by PengfeiM** [comment]
If I want to implement openvswitch on this. 
I need to install ovs and ryu for every server, right?

**Comment by bharnden** [comment]
yes, you need to have any software you plan to use installed on the actual host machine, for any given machine.

**Comment by PengfeiM** [comment]
Thanks for answering my question in the past days, 
I have a question, does CORE surppot **OVS & RYU** like EMANE so that I can just install them and use them with only one click in service configuration of the nodes. 
Or I need to write some configuration about **OVS & RYU**?

**Comment by bharnden** [comment]
Depends on what you are trying to do. There is an `--ovs` flag for the core-daemon that can allows the network to be linked by way of ovs bridges vs linux bridges if desired.

There are some basic OVS/RYU services provided, assuming they are installed. That provide some bare bones setup for one given take on setting things up.

If either of the above are good enough alone they may help. Otherwise you will need to write a custom service for your use case or orchestrate your setup after the scenario has started etc.

**Comment by PengfeiM** [comment]
Hello again, I installed ovs and also installed ryu following the document. 
But when I start the core by `sudo core-daemon --ovs` and `core-gui`, I can only see ovs in SDN service.  
what is this because?

**Comment by bharnden** [comment]
The Ryu service depends on having `ryu-manager` on PATH, so potentially that is your issue and the reason it is not showing up. This also means not your PATH as a user, but the PATH when CORE is ran under sudo/root.

`sudo which ryu-manager`

Also the `--ovs` flag is a different way of running CORE and not a requirement, you probably would not want to use that flag and try and use an OVS service on top of it.

**Comment by PengfeiM** [comment]
Thanks for that. 
Is there any document about ovs in CORE? 
Or where can I read the code about how ovs is  called by CORE in the source code.
Because currently, I can't get is work. 
Thanks. 
![image](https://user-images.githubusercontent.com/45191163/101114975-7f24c600-361d-11eb-9aea-227d2da65037.png)

**Comment by bharnden** [comment]
This is the example service contributed to CORE for one way to try and use OVS: https://github.com/coreemu/core/blob/a57b838f191640a49e45347a1764bf05b3269e0a/daemon/core/services/sdn.py#L24

This is by no means an end all solution to anyone wanting to use OVS, it is just one simple example that can be run. If you want things differently, you are getting to the point of looking into OVS documentation and potentially you would write a custom service to carry out your desired actions, or manually set things up by hand, etc.

At that point, this is not a CORE specific issue, but more of an issue for how you want to use OVS itself.
